File: algorithms/11_golang/strings/boyer_moore.go
// Boyer-Moore-Horspool substring search.
package main

import "fmt"

func boyerMoore(text, pat string) int {
	n, m := len(text), len(pat)
	skip := [256]int{}
	for i := range skip {
		skip[i] = m
	}
	for i := 0; i < m-1; i++ {
		skip[pat[i]] = m - 1 - i
	}
	i := 0
	for i+m <= n {
		j := m - 1
		for j >= 0 && text[i+j] == pat[j] {
			j--
		}
		if j < 0 {
			return i
		}
		i += skip[text[i+m-1]]
	}
	return -1
}

func main() {
	if boyerMoore("here is a simple example", "example") != 17 || boyerMoore("abc", "xyz") != -1 {
		panic("search failed")
	}
	fmt.Println("boyer-moore ok")
}

File: algorithms/13_swift/strings/boyer_moore.swift
// Boyer-Moore-Horspool substring search.
func boyerMoore(_ text: String, _ pattern: String) -> Int {
    let t = Array(text.utf8)
    let p = Array(pattern.utf8)
    let n = t.count
    let m = p.count
    var skip = [Int](repeating: m, count: 256)
    for i in 0..<(m - 1) { skip[Int(p[i])] = m - 1 - i }
    var i = 0
    while i + m <= n {
        var j = m - 1
        while j >= 0 && t[i + j] == p[j] { j -= 1 }
        if j < 0 { return i }
        i += skip[Int(t[i + m - 1])]
    }
    return -1
}

assert(boyerMoore("here is a simple example", "example") == 17)
assert(boyerMoore("abc", "xyz") == -1)
print("boyer-moore ok")
